external/storm-opentsdb/src/main/java/org/apache/stom/opentsdb/bolt/OpenTsdbBolt.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.storm.utils.BatchHelper;
+import org.apache.storm.utils.TupleUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Basic bolt implementation for storing timeseries datapoints to OpenTSDB.
+ *
+ * List of {@link OpenTsdbMetricDatapoint} is generated for each tuple with the configured {@code tupleOpenTsdbDatapointMappers}.
+ * All these datapoints are batched till the given {@code batchSize} or {@code flushIntervalInSeconds} is reached.
+ *
+ * Example topology:
+ *
+ */
+public class OpenTsdbBolt extends BaseRichBolt {
+    private static final Logger LOG = LoggerFactory.getLogger(OpenTsdbBolt.class);
+
+    private final OpenTsdbClient.Builder openTsdbClientBuilder;
+    private final List<TupleOpenTsdbDatapointMapper> tupleOpenTsdbDatapointMappers;
+    private int batchSize;
+    private int flushIntervalInSeconds;
+    private boolean failTupleForFailedMetrics;
+
+    private BatchHelper batchHelper;
+    private OpenTsdbClient openTsdbClient;
+    private Map<OpenTsdbMetricDatapoint, Tuple> metricPointsWithTuple = new HashMap<>();
+    private OutputCollector collector;
+
+    public OpenTsdbBolt(OpenTsdbClient.Builder openTsdbClientBuilder, List<TupleOpenTsdbDatapointMapper> tupleOpenTsdbDatapointMappers) {
+        this.openTsdbClientBuilder = openTsdbClientBuilder;
+        this.tupleOpenTsdbDatapointMappers = tupleOpenTsdbDatapointMappers;
+    }
+
+    public OpenTsdbBolt withFlushInterval(int flushIntervalInSeconds) {
+        this.flushIntervalInSeconds = flushIntervalInSeconds;
+        return this;
+    }
+
+    public OpenTsdbBolt withBatchSize(int batchSize) {
+        this.batchSize = batchSize;
+        return this;
+    }
+
+    /**
+     * When it is invoked, this bolt acks only the tuples which have successful metrics stored into OpenTSDB and fails
+     * the respective tuples of the failed metrics.
+     *
+     * @return same insatcne by setting {@code failTupleForFailedMetrics} to true
+     */
+    public OpenTsdbBolt failTupleForFailedMetrics() {
+        this.failTupleForFailedMetrics = true;
+        return this;
+    }
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        this.collector = collector;
+        batchHelper = new BatchHelper(batchSize, collector);
+        openTsdbClient = openTsdbClientBuilder.build();
+    }
+
+    @Override
+    public void execute(Tuple tuple) {
+        try {
+            if (batchHelper.shouldHandle(tuple)) {
+                final List<OpenTsdbMetricDatapoint> metricDataPoints = getMetricPoints(tuple);
+                for (OpenTsdbMetricDatapoint metricDataPoint : metricDataPoints) {
+                    metricPointsWithTuple.put(metricDataPoint, tuple);
+                }
+                batchHelper.addBatch(tuple);
```

Comment:
It just counts multiple metric datapoints into one since BatchHelper handles batch via tuple. In fact, "count of mappers" data points are created for each tuple.
Since target of batching is metric datapoints instead of tuples, it seems not intuitive behavior.